File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.isSame(root.left,root.right)
    def isSame(self,leftroot,rightroot):
        if leftroot == None and rightroot == None:
            return True
        elif leftroot == None or rightroot == None:
            return False
        elif leftroot.val != rightroot.val:
            return False
        else:
            return self.isSame(leftroot.left, rightroot.right) and self.isSame(leftroot.right, rightroot.left)
        
        

File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool check(TreeNode* l, TreeNode* r){
        //check if either of them is null
        if(l==nullptr || r ==nullptr){
            return l==r;
        }
        //if data doesnt match
        if(l->val != r->val){
            return false;
        }
        //check for both further sides simultaneously
        return check(l->left,r->right) && check(l->right,r->left);
    }
    bool isSymmetric(TreeNode* root) {
        if(root == nullptr){
            return true;
        }
        return check(root->left,root->right);
    }
};